### PR TITLE
Feat/aabb helpers

### PR DESCRIPTION
**Pull request by francoisbonneau**

### CODE REVIEW

**Comment by BotellaA** [comment]
One comment about commit names. Fix(XXX), XXX should be as short as possible and close to a category like in your case "Distance" would be enough.

You did not add à commit for the aabb helper itself

File: include/geode/mesh/core/aabb_helpers.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+/*
+ * Modified from RINGMesh https://github.com/ringmesh/RINGMesh
+ * Copyright (c) 2012-2018, Association Scientifique pour la Geologie et ses
+ * Applications (ASGA)
+ */
+
+#pragma once
+
+#include <geode/basic/pimpl.h>
```
Comment (BotellaA) [inline include/geode/mesh/core/aabb_helpers.h:32]: not used

File: include/geode/mesh/core/aabb_helpers.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+#include <geode/geometry/aabb.h>
+#include <geode/geometry/point.h>
```
Comment (BotellaA) [inline include/geode/mesh/core/aabb_helpers.h:35]: forward declaration should be enough

File: include/geode/mesh/core/aabb_helpers.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    public:
+        explicit DistanceToTriangle(
+            const TriangulatedSurface< dimension >& mesh )
+            : mesh_( mesh )
+        {
+        }
+
+        std::tuple< double, Point< dimension > > operator()(
+            const Point< dimension >& query, index_t cur_box ) const;
+        double operator()( const Point< dimension >& pt1,
```
Comment (BotellaA) [inline include/geode/mesh/core/aabb_helpers.h:62]: I would prefer more complete names such as point1 and point2

File: include/geode/mesh/core/aabb_helpers.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace geode
+{
+    FORWARD_DECLARATION_DIMENSION_CLASS( TriangulatedSurface );
+} // namespace geode
+
+namespace geode
+{
+    template < index_t dimension >
+    std::unique_ptr< AABBTree< dimension > > get_aabb(
```
Comment (BotellaA) [inline include/geode/mesh/core/aabb_helpers.h:47]: `get_aabb` is not really what is happening here, you are creating a new AABBTree and not getting it from somewhere. Maybe `create_aabb_tree`? @panquez an idea of a good name?

Other point, can you return directly an AABBTree instead of a unique_ptr?

File: src/geode/geometry/distance.cpp [hunk 5]
```diff
@@ -299,43 +316,32 @@ namespace geode
...
-        double result;
-        Point2D closest_point;
-        if( distance[0] < distance[1] )
+        auto dist = MAX_DOUBLE;
+        Point2D nearest_p;
+        bool inside{ true };
+        for( const auto s : Range{ 3 } )
         {
-            if( distance[0] < distance[2] )
+            auto distance = MAX_DOUBLE;
```
Comment (BotellaA) [inline src/geode/geometry/distance.cpp:324]:
```suggestion
            double distance;
```

File: src/geode/geometry/distance.cpp [hunk 6]
```diff
@@ -299,43 +316,32 @@ namespace geode
...
-        std::tie( distance[0], closest[0] ) = point_segment_distance( point,
-            Segment2D{ triangle.vertices()[0], triangle.vertices()[1] } );
-        std::tie( distance[1], closest[1] ) = point_segment_distance( point,
-            Segment2D{ triangle.vertices()[1], triangle.vertices()[2] } );
-        std::tie( distance[2], closest[2] ) = point_segment_distance( point,
-            Segment2D{ triangle.vertices()[2], triangle.vertices()[0] } );
-        double result;
-        Point2D closest_point;
-        if( distance[0] < distance[1] )
+        auto dist = MAX_DOUBLE;
```
Comment (BotellaA) [inline src/geode/geometry/distance.cpp:319]: maybe rename it as `max_distance`

File: src/geode/geometry/distance.cpp [hunk 7]
```diff
@@ -299,43 +316,32 @@ namespace geode
...
-                result = distance[1];
-                closest_point = closest[1];
-            }
-            else
-            {
-                result = distance[2];
-                closest_point = closest[2];
-            }
+            nearest_p = point;
+            return std::make_tuple( 0.0, nearest_p );
```
Comment (BotellaA) [inline src/geode/geometry/distance.cpp:342]:
```suggestion
            return std::make_tuple( 0.0, point );
```

File: src/geode/mesh/core/aabb_helpers.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+ */
+
+#include <geode/mesh/core/aabb_helpers.h>
+#include <geode/mesh/core/triangulated_surface.h>
+
+#include <geode/geometry/basic_objects.h>
+#include <geode/geometry/distance.h>
+#include <geode/geometry/vector.h>
+
+#include <memory>
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_helpers.cpp:37]:
```suggestion
#include <geode/mesh/core/aabb_helpers.h>

#include <memory>

#include <geode/geometry/basic_objects.h>
#include <geode/geometry/distance.h>
#include <geode/geometry/vector.h>

#include <geode/mesh/core/triangulated_surface.h>
```
Include order:

1. corresponding header
1. external dependencies (std, then abseil, ...)
1. one block per library, still orderer by number of dependencies

File: src/geode/mesh/core/aabb_helpers.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace geode
+{
+    template < index_t dimension >
+    std::unique_ptr< AABBTree< dimension > > get_aabb(
+        const TriangulatedSurface< dimension >& mesh )
+    {
+        std::vector< BoundingBox< dimension > > box_vector(
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_helpers.cpp:45]:
```suggestion
        absl::FixedArray< BoundingBox< dimension > > box_vector(
```
When you know the number of elements at runtime, prefer absl::FixedArray.

File: src/geode/mesh/core/aabb_helpers.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            mesh.nb_polygons() );
+        for( auto p : Range( mesh.nb_polygons() ) )
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_helpers.cpp:47]:
```suggestion
        for( const auto p : Range{ mesh.nb_polygons() } )
```

File: src/geode/mesh/core/aabb_helpers.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            for( auto v : Range( mesh.nb_polygon_vertices( p ) ) )
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_helpers.cpp:49]:
```suggestion
            for( const auto v : Range{ mesh.nb_polygon_vertices( p ) } )
```

File: src/geode/mesh/core/aabb_helpers.cpp [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+                    mesh.point( mesh.polygon_vertex( { p, v } ) ) );
+            }
+        }
+        return std::unique_ptr< AABBTree< dimension > >(
+            new AABBTree< dimension >( box_vector ) );
+    }
+    template opengeode_mesh_api std::unique_ptr< AABBTree2D > get_aabb< 2 >(
+        const TriangulatedSurface2D& );
+    template opengeode_mesh_api std::unique_ptr< AABBTree3D > get_aabb< 3 >(
+        const TriangulatedSurface3D& );
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_helpers.cpp:61]: Is there a reason to put the specializations here and not at the bottom of the file?

File: src/geode/mesh/core/aabb_helpers.cpp [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        return point_triangle_distance(
+            query, Triangle< dimension >{ v0, v1, v2 } );
+    }
+
+    template < index_t dimension >
+    double DistanceToTriangle< dimension >::operator()(
+        const Point< dimension >& pt1, const Point< dimension >& pt2 ) const
+    {
+        const Vector< dimension > vec{ pt1, pt2 };
+        return vec.length();
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_helpers.cpp:79]:
```suggestion
        return Vector< dimension >{ pt1, pt2 }.length();
```

File: tests/mesh/test-aabb-helpers.cpp [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+            builder.set_point( id++, create_vertex< dimension >( i, j ) );
+        }
+    }
+}
+
+template < index_t dimension >
+void add_triangles(
+    TriangulatedSurfaceBuilder< dimension >& builder, index_t size )
+{
+    builder.create_triangles( ( size - 1 ) * ( size - 1 ) * 2 );
```
Comment (BotellaA) [inline tests/mesh/test-aabb-helpers.cpp:69]: prefer `reserve_triangles`, `set_polygon_vertex` is slower than creating it directly.

File: tests/mesh/test-aabb-helpers.cpp [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    index_t id = 0;
+    for( index_t i : Range( size - 1 ) )
```
Comment (BotellaA) [inline tests/mesh/test-aabb-helpers.cpp:71]:
```suggestion
    for( const auto i : Range{ size - 1 } )
```

File: tests/mesh/test-aabb-helpers.cpp [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+#include <geode/basic/logger.h>
+
+#include <geode/mesh/builder/triangulated_surface_builder.h>
+#include <geode/mesh/core/triangulated_surface.h>
+
+#include <geode/mesh/core/aabb_helpers.h>
+
+#include <geode/tests/common.h>
+
+using namespace geode;
```
Comment (BotellaA) [inline tests/mesh/test-aabb-helpers.cpp:39]: We prefer not using this.

File: tests/mesh/test-aabb-helpers.cpp [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+
+template < index_t dimension >
+void add_triangles(
+    TriangulatedSurfaceBuilder< dimension >& builder, index_t size )
+{
+    builder.create_triangles( ( size - 1 ) * ( size - 1 ) * 2 );
+    index_t id = 0;
+    for( index_t i : Range( size - 1 ) )
+    {
+        for( index_t j : Range( size - 1 ) )
```
Comment (BotellaA) [inline tests/mesh/test-aabb-helpers.cpp:73]: same here

File: tests/mesh/test-aabb-helpers.cpp [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+template < index_t dimension >
+void check_tree( const AABBTree< dimension >& tree,
+    const DistanceToTriangle< dimension >& distance_action,
+    index_t size )
+{
+    double offset = 0.2;
+    index_t id = 0;
+    for( index_t i : Range( size - 1 ) )
+    {
+        for( index_t j : Range( size - 1 ) )
```
Comment (BotellaA) [inline tests/mesh/test-aabb-helpers.cpp:96]: same iteration convention than before

File: tests/mesh/test-aabb-helpers.cpp [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+        }
+    }
+}
+
+template < index_t dimension >
+void check_tree( const AABBTree< dimension >& tree,
+    const DistanceToTriangle< dimension >& distance_action,
+    index_t size )
+{
+    double offset = 0.2;
```
Comment (BotellaA) [inline tests/mesh/test-aabb-helpers.cpp:92]:
```suggestion
    constexpr auto offset = 0.2;
```

File: tests/mesh/test-aabb-helpers.cpp [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+    index_t id = 0;
+    for( index_t i : Range( size - 1 ) )
+    {
+        for( index_t j : Range( size - 1 ) )
+        {
+            Point< dimension > query1 =
```
Comment (BotellaA) [inline tests/mesh/test-aabb-helpers.cpp:98]:
```suggestion
            const auto query1 =
```
prefer always `const auto` if possible

File: tests/mesh/test-aabb-helpers.cpp [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+                create_vertex< dimension >( i + offset, j + offset );
+            index_t triangle1 = NO_ID;
```
Comment (BotellaA) [inline tests/mesh/test-aabb-helpers.cpp:100]: no need to initialize these values if you use then in tie

File: tests/mesh/test-aabb-helpers.cpp (hunk 20 above)
Comment (BotellaA) [reply tests/mesh/test-aabb-helpers.cpp:98]: Could you do it for the entire file?

File: include/geode/mesh/core/aabb_helpers.h (hunk 4 above)
Comment (panquez) [reply include/geode/mesh/core/aabb_helpers.h:47]: I would say `create_aabb_tree` too since you create it.

File: src/geode/geometry/distance.cpp [hunk 22]
```diff
@@ -45,6 +46,22 @@ namespace geode
...
+        const Point2D& point, const Segment2D& segment )
+    {
+        Point2D nearest_point;
+        double distance;
+        std::tie( distance, nearest_point ) =
+            point_segment_distance< 2 >( point, segment );
+        const Vector2D proj2point{ nearest_point, point };
+        // segment 2D facet normals point towards inside
+        const auto signed_distance =
+            dot_perpendicular( proj2point, segment.direction() ) <= 0
```
Comment (panquez) [inline src/geode/geometry/distance.cpp:59]: should we use `dot_perpendicular` or `triangle_signed_area` using query and segment extremities? For info: `dot_perpendicular` is used in `triangle_signed_area`: 
```        return dot_perpendicular(
                   { triangle.vertices()[0], triangle.vertices()[1] },
                   { triangle.vertices()[0], triangle.vertices()[2] } )
               / 2.;
```
It could prevent being unhomogoneous.

File: src/geode/geometry/distance.cpp [hunk 23]
```diff
@@ -45,6 +46,22 @@ namespace geode
...
 
+    std::tuple< double, Point2D > point_segment_signed_distance(
+        const Point2D& point, const Segment2D& segment )
+    {
+        Point2D nearest_point;
+        double distance;
+        std::tie( distance, nearest_point ) =
+            point_segment_distance< 2 >( point, segment );
+        const Vector2D proj2point{ nearest_point, point };
+        // segment 2D facet normals point towards inside
```
Comment (panquez) [inline src/geode/geometry/distance.cpp:57]: I do not understand  this comment? What is a "segment facet normals" ?

File: src/geode/mesh/core/aabb_helpers.cpp [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+        for( auto p : Range( mesh.nb_polygons() ) )
+        {
+            for( auto v : Range( mesh.nb_polygon_vertices( p ) ) )
+            {
+                box_vector[p].add_point(
+                    mesh.point( mesh.polygon_vertex( { p, v } ) ) );
+            }
+        }
+        return std::unique_ptr< AABBTree< dimension > >(
+            new AABBTree< dimension >( box_vector ) );
```
Comment (panquez) [inline src/geode/mesh/core/aabb_helpers.cpp:56]: Why not returning directly the AABBTree?

File: tests/geometry/test-distance.cpp [hunk 25]
```diff
@@ -158,8 +158,8 @@ void test_point_segment_distance()
 void test_point_triangle_distance_2d()
 {
     const geode::Point2D a{ { 0.0, 0.0 } };
-    const geode::Point2D b{ { 1.0, 0.0 } };
-    const geode::Point2D c{ { 1.0, 1.0 } };
+    const geode::Point2D b{ { 1.0, 1.0 } };
+    const geode::Point2D c{ { 1.0, 0.0 } };
```
Comment (panquez) [inline tests/geometry/test-distance.cpp:162]: This change is OK for me, but it is a bit strange since you may break other tests based on previous configuration.

File: tests/geometry/test-distance.cpp [hunk 26]
```diff
@@ -177,6 +177,13 @@ void test_point_triangle_distance_2d()
         "[Test] Wrong result for point_triangle_distance with query point "
         "b" );
 
+    const geode::Point2D q0{ { 0.8, 0.2 } };
+    std::tie( distance, closest_point ) =
+        geode::point_triangle_distance( q0, triangle2D );
+    OPENGEODE_EXCEPTION( distance == 0 && closest_point == q0,
+        "[Test] Wrong result for point_triangle_distance with query point "
+        "q0" );
+
```
Comment (panquez) [inline tests/geometry/test-distance.cpp:186]: Could you remove q4 case that is a duplication of q1 case?

File: tests/mesh/test-aabb-helpers.cpp (hunk 16 above)
Comment (panquez) [reply tests/mesh/test-aabb-helpers.cpp:39]: I agree... even if I think this is a lot of this in our tests :/

File: tests/mesh/test-aabb-helpers.cpp [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+
+template < index_t dimension >
+void add_vertices(
+    TriangulatedSurfaceBuilder< dimension >& builder, index_t size )
+{
+    builder.create_vertices( size * size );
+    index_t id = 0;
+    for( index_t i : Range( size ) )
+    {
+        for( index_t j : Range( size ) )
```
Comment (panquez) [inline tests/mesh/test-aabb-helpers.cpp:58]: Replace index_t by `const auto`, you can also replace Range() by Range{}

File: src/geode/geometry/distance.cpp (hunk 22 above)
Comment (BotellaA) [reply src/geode/geometry/distance.cpp:59]: True

File: src/geode/mesh/core/aabb_helpers.cpp (hunk 24 above)
Comment (BotellaA) [reply src/geode/mesh/core/aabb_helpers.cpp:56]: True, need a move constructor for the tree

File: src/geode/geometry/distance.cpp (hunk 22 above)
Comment (panquez) [reply src/geode/geometry/distance.cpp:59]: I would use query point as Triangle vertex 0 and segment extremities being the two other vertices in the same order

File: include/geode/mesh/core/aabb_helpers.h (hunk 1 above)
Comment (francoisbonneau) [reply include/geode/mesh/core/aabb_helpers.h:32]: yes should we used a pimpl?

Comment (panquez) [reply include/geode/mesh/core/aabb_helpers.h:32]: I think no since it is just a API function. Maybe for the class, but I would say no too since this class is (1) very little, (2) it will be used as functor and as example for doing other functors (3) there is nothing to hide in terms of implementation

File: src/geode/geometry/distance.cpp [hunk 28]
```diff
@@ -26,12 +26,30 @@
...
 #include <geode/geometry/vector.h>
 
 namespace
 {
     constexpr double MAX_DOUBLE = std::numeric_limits< double >::max();
+
+    bool point_is_in_triangle(
+        const geode::Point2D& point, const geode::Triangle2D& triangle )
+    {
+        double signed_area_1 = geode::triangle_signed_area( geode::Triangle2D{
```
Comment (BotellaA) [inline src/geode/geometry/distance.cpp:41]: Review the file and prefer const auto when possible

File: src/geode/mesh/core/aabb_helpers.cpp [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+        absl::FixedArray< BoundingBox< dimension > > box_vector(
+            mesh.nb_edges() );
+        for( const auto e : Range{ mesh.nb_edges() } )
+        {
+            box_vector[e].add_point(
+                mesh.point( mesh.edge_vertex( { e, 0 } ) ) );
+            box_vector[e].add_point(
+                mesh.point( mesh.edge_vertex( { e, 1 } ) ) );
+        }
+        return AABBTree< dimension >( box_vector );
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_helpers.cpp:58]: Use {} when creating an object as much as possible. Same comment at several location in the file.

Comment (francoisbonneau) [reply src/geode/mesh/core/aabb_helpers.cpp:58]: should I write         
absl::FixedArray< BoundingBox< dimension > > box_vector{
            mesh.nb_edges()
        };
as well... I do not like the way clang is formatting this

File: include/geode/mesh/core/aabb_edge_curve_helpers.h [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+    public:
+        explicit DistanceToEdge( const EdgedCurve< dimension >& mesh )
+            : mesh_( mesh )
+        {
+        }
+
+        std::tuple< double, Point< dimension > > operator()(
+            const Point< dimension >& query, index_t cur_box ) const;
+        double operator()( const Point< dimension >& point1,
+            const Point< dimension >& point2 ) const;
```
Comment (BotellaA) [inline include/geode/mesh/core/aabb_edge_curve_helpers.h:59]: Do you still need both?

File: include/geode/mesh/core/aabb_triangulated_surface_helpers.h [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+            const TriangulatedSurface< dimension >& mesh )
+            : mesh_( mesh )
+        {
+        }
+
+        std::tuple< double, Point< dimension > > operator()(
+            const Point< dimension >& query, index_t cur_box ) const;
+        double operator()( const Point< dimension >& point1,
+            const Point< dimension >& point2 ) const;
+
```
Comment (BotellaA) [inline include/geode/mesh/core/aabb_triangulated_surface_helpers.h:61]: same question

File: src/geode/mesh/core/aabb_triangulated_surface_helpers.cpp [hunk 32]
```diff
@@ -0,0 +1,88 @@
...
+/*
+ * Modified from RINGMesh https://github.com/ringmesh/RINGMesh
+ * Copyright (c) 2012-2018, Association Scientifique pour la Geologie et ses
+ * Applications (ASGA)
+ */
+
+#include <geode/mesh/core/aabb_triangulated_surface_helpers.h>
+
+#include <memory>
+
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_triangulated_surface_helpers.cpp:33]:
```suggestion
```

File: src/geode/mesh/core/aabb_edge_curve_helpers.cpp [hunk 33]
```diff
@@ -0,0 +1,84 @@
...
+/*
+ * Modified from RINGMesh https://github.com/ringmesh/RINGMesh
+ * Copyright (c) 2012-2018, Association Scientifique pour la Geologie et ses
+ * Applications (ASGA)
+ */
+
+#include <geode/mesh/core/aabb_edge_curve_helpers.h>
+
+#include <memory>
+
```
Comment (BotellaA) [inline src/geode/mesh/core/aabb_edge_curve_helpers.cpp:33]:
```suggestion
```

File: include/geode/mesh/core/aabb_edge_curve_helpers.h (hunk 30 above)
Comment (francoisbonneau) [reply include/geode/mesh/core/aabb_edge_curve_helpers.h:59]: good point I forgot our modificatio

File: src/geode/geometry/distance.cpp [hunk 34]
```diff
@@ -26,12 +26,33 @@
...
+#include <geode/geometry/perpendicular.h>
 #include <geode/geometry/projection.h>
+#include <geode/geometry/signed_mensuration.h>
 #include <geode/geometry/vector.h>
 
 namespace
 {
     constexpr double MAX_DOUBLE = std::numeric_limits< double >::max();
+
+    bool point_is_in_triangle(
```
Comment (BotellaA) [inline src/geode/geometry/distance.cpp:38]: Should we rename this in something with approx in the name? because this function should not be used for exact answer. How about `may_point_is_in_triangle`? @panquez what do you think?

File: tests/mesh/test-aabb-triangulated-surface-helpers.cpp [hunk 35]
```diff
@@ -0,0 +1,156 @@
...
+        }
+    }
+}
+
+template < geode::index_t dimension >
+void add_triangles( geode::TriangulatedSurfaceBuilder< dimension >& builder,
+    geode::index_t size )
+{
+    builder.reserve_triangles( ( size - 1 ) * ( size - 1 ) * 2 );
+    geode::index_t id = 0;
```
Comment (panquez) [inline tests/mesh/test-aabb-triangulated-surface-helpers.cpp:70]: you can remove id (unused)

File: tests/mesh/test-aabb-triangulated-surface-helpers.cpp [hunk 36]
```diff
@@ -0,0 +1,156 @@
...
+
+    auto t_surf = geode::TriangulatedSurface< dimension >::create();
+    auto t_surf_builder =
+        geode::TriangulatedSurfaceBuilder< dimension >::create( *t_surf );
+
+    geode::index_t size = 10;
+    add_vertices( *t_surf_builder, size );
+    add_triangles( *t_surf_builder, size );
+
+    auto aabb_tree = create_aabb_tree( *t_surf );
```
Comment (panquez) [inline tests/mesh/test-aabb-triangulated-surface-helpers.cpp:144]: const?

File: tests/mesh/test-aabb-edge-curve-helpers.cpp [hunk 37]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        builder.set_edge_vertex( { i, 0 }, i );
+        builder.set_edge_vertex( { i, 1 }, i + 1 );
+    }
+}
+
+template < geode::index_t dimension >
+void check_edgecurve_tree( const geode::EdgedCurve< dimension >& edge_curve )
+{
+    auto aabb_tree = geode::create_aabb_tree< dimension >( edge_curve );
```
Comment (panquez) [inline tests/mesh/test-aabb-edge-curve-helpers.cpp:76]: const

File: tests/mesh/test-aabb-edge-curve-helpers.cpp [hunk 38]
```diff
@@ -0,0 +1,111 @@
...
+    geode::DistanceToEdge< dimension > distance_action( edge_curve );
```
Comment (panquez) [inline tests/mesh/test-aabb-edge-curve-helpers.cpp:77]: const?

File: src/geode/geometry/distance.cpp (hunk 34 above)
Comment (BotellaA) [reply src/geode/geometry/distance.cpp:38]: am not sure, people do not read comment but they have to read the function name :p

Comment (francoisbonneau) [reply src/geode/geometry/distance.cpp:38]: in any case the function is not exported

File: src/geode/geometry/distance.cpp [hunk 39]
```diff
@@ -35,7 +35,12 @@ namespace
 {
     constexpr double MAX_DOUBLE = std::numeric_limits< double >::max();
 
-    bool point_is_in_triangle(
+    /*
+     * This function test if the point is in the triangle by computing signed
+     * areas
+     * @warning This is not robust in any cases!
```
Comment (BotellaA) [inline src/geode/geometry/distance.cpp:41]:
```suggestion
     * @warning This is not robust in edge cases!
```

File: src/geode/geometry/distance.cpp [hunk 40]
```diff
@@ -26,12 +26,38 @@
...
 #include <geode/geometry/projection.h>
+#include <geode/geometry/signed_mensuration.h>
 #include <geode/geometry/vector.h>
 
 namespace
 {
     constexpr double MAX_DOUBLE = std::numeric_limits< double >::max();
+
+    /*
+     * This function test if the point is in the triangle by computing signed
```
Comment (panquez) [inline src/geode/geometry/distance.cpp:39]:
```suggestion
     * This function tests if the point is in the triangle by computing signed
```

File: src/geode/geometry/distance.cpp [hunk 41]
```diff
@@ -26,12 +26,38 @@
...
+     * areas
+     * @warning This is not robust in edge cases!
+     */
+    bool may_point_is_in_triangle(
```
Comment (panquez) [inline src/geode/geometry/distance.cpp:43]:
```suggestion
    bool may_point_be_in_triangle(
```

File: src/geode/geometry/distance.cpp [hunk 42]
```diff
@@ -299,6 +339,10 @@ namespace geode
     std::tuple< double, Point2D > point_triangle_distance(
         const Point2D& point, const Triangle2D& triangle )
     {
+        if( may_point_is_in_triangle( point, triangle ) )
```
Comment (panquez) [inline src/geode/geometry/distance.cpp:342]:
```suggestion
        if( may_point_be_in_triangle( point, triangle ) )
```

File: tests/mesh/test-aabb-edge-curve-helpers.cpp [hunk 43]
```diff
@@ -0,0 +1,111 @@
...
+    check_edgecurve_tree< dimension >( *edge_curve );
+}
+
+void test()
+{
+    test_EdgeCurveAABB< 2 >();
+    test_EdgeCurveAABB< 3 >();
+}
+
+OPENGEODE_TEST( "aabb-edge-curve-helpers" )
```
Comment (panquez) [inline tests/mesh/test-aabb-edge-curve-helpers.cpp:111]: EdgedCurve. Could you rename the files and correct the typo?

File: tests/mesh/test-aabb-triangulated-surface-helpers.cpp [hunk 44]
```diff
@@ -0,0 +1,153 @@
...
+    }
+}
+
+template < geode::index_t dimension >
+void check_surface_tree( const geode::AABBTree< dimension >& tree,
+    const geode::DistanceToTriangle< dimension >& distance_action,
+    geode::index_t size )
+{
+    constexpr auto offset = 0.2;
+    geode::index_t id = 0;
```
Comment (panquez) [inline tests/mesh/test-aabb-triangulated-surface-helpers.cpp:88]:
```suggestion
    geode::index_t id{ 0 };
```

File: tests/mesh/test-aabb-triangulated-surface-helpers.cpp [hunk 45]
```diff
@@ -0,0 +1,153 @@
...
+void test_SurfaceAABB()
+{
+    geode::Logger::info(
+        "TEST", " TriangulatedSurface AABB Helper", dimension, "D" );
+
+    auto t_surf = geode::TriangulatedSurface< dimension >::create();
+    auto t_surf_builder =
+        geode::TriangulatedSurfaceBuilder< dimension >::create( *t_surf );
+
+    geode::index_t size = 10;
```
Comment (panquez) [inline tests/mesh/test-aabb-triangulated-surface-helpers.cpp:137]:
```suggestion
    geode::index_t size{ 10 };
```
